core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/index/IndexVariableListener.java

```diff
@@ -37,81 +37,87 @@ public IndexVariableListener(
+        // Do nothing
     }
 
     @Override
-    public void beforeEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+    public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        updateIndexes((InnerScoreDirector<Solution_, ?>) scoreDirector, entity, 0);
+    }
+
+    @Override
+    public void beforeEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
         // Do nothing
     }
 
     @Override
-    public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object entity) {
-        insert((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    public void afterEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        List<Object> listVariable = sourceVariableDescriptor.getListVariable(entity);
+        for (Object element : listVariable) {
+            innerScoreDirector.beforeVariableChanged(shadowVariableDescriptor, element);
+            shadowVariableDescriptor.setValue(element, null);
+            innerScoreDirector.afterVariableChanged(shadowVariableDescriptor, element);
+        }
     }
 
     @Override
-    public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
-        retract((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    public void beforeElementAdded(ScoreDirector<Solution_> scoreDirector, Object entity, int index) {
+        // Do nothing
     }
 
     @Override
-    public void afterVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
-        insert((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    public void afterElementAdded(ScoreDirector<Solution_> scoreDirector, Object entity, int index) {
+        updateIndexes((InnerScoreDirector<Solution_, ?>) scoreDirector, entity, index);
     }
 
     @Override
-    public void beforeEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
-        retract((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
+    public void beforeElementRemoved(ScoreDirector<Solution_> scoreDirector, Object entity, int index) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        Object element = sourceVariableDescriptor.getElement(entity, index);
+        innerScoreDirector.beforeVariableChanged(shadowVariableDescriptor, element);
+        shadowVariableDescriptor.setValue(element, null);
+        innerScoreDirector.afterVariableChanged(shadowVariableDescriptor, element);
     }
 
     @Override
-    public void afterEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object entity) {
+    public void afterElementRemoved(ScoreDirector<Solution_> scoreDirector, Object entity, int index) {
+        updateIndexes((InnerScoreDirector<Solution_, ?>) scoreDirector, entity, index);
```

Comment:
The purpose of before methods is to offer a chance to listeners to look at the state of the solution before a change is performed but they're not supposed to make any changes unless necessary. Most of the `before*` overrides do nothing.

In case of the `elementRemoved` event, which is used by the unassign move, the after method is too late to unset the removed element's shadow variable because the removed element is no longer at its original place (entity[index]).

What we could do is to remember the element that's being removed and defer the update of its shadow variable to the after method. But that would make the listener stateful, which introduces more complexity and error proneness. That's why we're making and exception to the rule and we unset the removed element's shadow variable in the before method.

To minimize the changes done in the before method, the remaining elements in the modified entity's list variable are updated in the after method.